File: src/main/scala/fr/xebia/streams/LocalWebcamWindow.scala
package fr.xebia.streams

import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import fr.xebia.streams.common.Dimensions
import fr.xebia.streams.transform.{Flip, MediaConversion}
import fr.xebia.streams.video.ImageProcessingSinks.ShowImageSink
import fr.xebia.streams.video.Webcam
import org.bytedeco.javacv.CanvasFrame

object LocalWebcamWindow extends App {

  implicit val system = ActorSystem()
  implicit val materializer = ActorMaterializer()

  val canvas = new CanvasFrame("Webcam")
  //  //Set Canvas frame to close on exit
  canvas.setDefaultCloseOperation(javax.swing.JFrame.EXIT_ON_CLOSE)

  val imageDimensions = Dimensions(width = 640, height = 480)
  val localCameraSource = Webcam.local(deviceId = 0, dimensions = imageDimensions)

  val graph = localCameraSource
    .map(MediaConversion.frameToMat) // most OpenCV manipulations require a Matrix
    .map(Flip.horizontal)
    .map(MediaConversion.matToFrame) // convert back to a frame
    .to(ShowImageSink(canvas))

  graph.run()

}

File: src/main/java/fr/xebia/streams/java/LocalWebCamWindow.java
package fr.xebia.streams.java;

import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.stream.ActorMaterializer;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import fr.xebia.streams.java.common.Dimensions;
import fr.xebia.streams.java.video.Local;
import fr.xebia.streams.transform.Flip;
import fr.xebia.streams.transform.MediaConversion;
import org.bytedeco.javacv.CanvasFrame;
import org.bytedeco.javacv.Frame;
import org.bytedeco.javacv.FrameGrabber;

import static org.bytedeco.javacpp.opencv_core.CV_8U;

public class LocalWebCamWindow {

    public void init() {
        CanvasFrame webcam = new CanvasFrame("Webcam");
        webcam.setDefaultCloseOperation(javax.swing.JFrame.EXIT_ON_CLOSE);

        Dimensions imageDimensions = new Dimensions(640, 480);
        ActorSystem actorSystem = ActorSystem.create();
        ActorMaterializer materializer = ActorMaterializer.create(actorSystem);
        Source<Frame, NotUsed> localCameraSource = Local.localWebCam.apply(0, imageDimensions, CV_8U, FrameGrabber.ImageMode.COLOR, actorSystem);

        localCameraSource
                .map(MediaConversion::toMat) // most OpenCV manipulations require a Matrix
                .map(Flip::horizontal)
                .map(MediaConversion::toFrame) // convert back to a frame
                .map((image) -> {
                    webcam.showImage(image);
                    return image;
                })
                .to(Sink.ignore())
                .run(materializer);
    }


    public static void main(String args[]){
        new LocalWebCamWindow().init();

    }


}
